src/Compilers/CSharp/Portable/Symbols/Source/LocalFunctionSymbol.cs

```diff
@@ -423,33 +423,49 @@ private ImmutableArray<SourceMethodTypeParameterSymbol> MakeTypeParameters(Diagn
             return result.ToImmutableAndFree();
         }
 
-        public override ImmutableArray<TypeParameterConstraintClause> TypeParameterConstraintClauses
+        public override ImmutableArray<TypeParameterConstraintClause> GetTypeParameterConstraintClauses(bool early)
         {
-            get
+            var clauses = _lazyTypeParameterConstraints;
+            if (clauses.IsDefault)
             {
-                if (_lazyTypeParameterConstraints.IsDefault)
+                // Early step.
+                var diagnostics = DiagnosticBag.GetInstance();
+                var constraints = this.MakeTypeParameterConstraintsEarly(
+                    _binder,
+                    TypeParameters,
+                    _syntax.ConstraintClauses,
+                    _syntax.Identifier.GetLocation(),
+                    diagnostics);
+                lock (_declarationDiagnostics)
                 {
-                    var diagnostics = DiagnosticBag.GetInstance();
-                    var constraints = this.MakeTypeParameterConstraints(
-                        _binder,
-                        TypeParameters,
-                        _syntax.ConstraintClauses,
-                        _syntax.Identifier.GetLocation(),
-                        diagnostics);
-
-                    lock (_declarationDiagnostics)
+                    if (_lazyTypeParameterConstraints.IsDefault)
                     {
-                        if (_lazyTypeParameterConstraints.IsDefault)
-                        {
-                            _declarationDiagnostics.AddRange(diagnostics);
-                            _lazyTypeParameterConstraints = constraints;
-                        }
+                        _declarationDiagnostics.AddRange(diagnostics);
+                        _lazyTypeParameterConstraints = constraints;
                     }
-                    diagnostics.Free();
                 }
+                diagnostics.Free();
+                clauses = _lazyTypeParameterConstraints;
+            }
 
-                return _lazyTypeParameterConstraints;
+            if (!early && clauses.IsEarly())
+            {
+                // Late step.
+                var diagnostics = DiagnosticBag.GetInstance();
+                var constraints = this.MakeTypeParameterConstraintsLate(TypeParameters, clauses, diagnostics);
```

Comment:
Same suggestion on the check to `_lazyTypeParameterConstraints.IsEarly()` to avoid unnecessary lock contention.